### ISSUE TITLE
Option to exclude check from some file

### ISSUE THREAD

**Issue by Soullivaneuh** [maglnet/ComposerRequireChecker#58]
Ref: Report https://github.com/maglnet/ComposerRequireChecker/issues/55#issuecomment-371168782

It would be great to be able to specify a file where some symbol check will be ignored.

We can simply provide a list of file to totally exclude:

```json
{
  "exclude-files" : [
    "config/bundles.php",
    "dummy/other/magic_file.php",
  ]
}
```

Or also provide the namespace to ignore on each file:

```json
{
  "exclude-files" : [
    "config/bundles.php": [
        "Some\\Namespace\\To\\Ignore",
    ],
    "dummy/other/magic_file.php",
  ]
}
```

As the original issue was about dev dependencies, the best would be also to specified when to also use dev dependencies while checking a file:

```json
{
  "check-dev-files" : [
    "config/bundles.php",
    "dummy/other/magic_file.php",
  ]
}
```

Note: The option name can/must be improved, it is just an illustration for now.

**Comment by mpdude** [comment]
I know that Symfony best practices are outside the scope of this project. However, for a long time there has been a practice of keeping tests inside `src/Some/Bundle/Tests/*`, and so it would be valuable if we could exclude these directories from checks.

**Comment by Ocramius** [comment]
I think this is fixed by https://github.com/symfony/symfony/pull/33579

**Comment by mpdude** [comment]
Not sure why... does it?

**Comment by Ocramius** [comment]
It basically drops rhose tests from the (confusing) source tree of symfony components.

In practice, symfony could split the sources from the tests, which would be a better fix, but the above suffices.

**Comment by mpdude** [comment]
Sorry for the noise... What I meant was that in many projects that I know, there is `src/App/whatever/Tests`, because this has been the place suggested for (project-specific) tests in the past.

I think Symfony best practices nowadays suggest something else 🙏🏻. But that makes it hard to convince people to use this tool here if there is no simple way to make it pass on CI as well. 

If we could exclude `src/App/whatever/**/Tests`, things might be easily fixable (just add the dependencies that were really missing).

**Comment by Ocramius** [comment]
@mpdude well, they'd need to be declared as "shadowed" in `composer.json` somehow.

We should most certainly not add more conventions/configs than what is already very clear from `composer.json`: moving files is trivial and takes minutes anyway.

**Comment by VincentLanglet** [comment]
In my case, the `.phpstorm.meta.php` are checked by this library and returning multiple unknown symbol. I would prefer to remove these files from the check.

**Comment by Ocramius** [comment]
That depends on your autoloader configuration

**Comment by VincentLanglet** [comment]
I have
```
"autoload": {
        "psr-4": {
            "Foo\\": "src/"
        }
    },
```

You mean I can easily exclude the phpstorm meta from the composer autoload and still having in the src folder ?

**Comment by Ocramius** [comment]
That file doesn't need to be in src in first place

**Comment by VincentLanglet** [comment]
> That file doesn't need to be in src in first place

I agree, it doesn't need to. But I have multiple `.phpstorm.meta.php` configuring different functions, different classes, ... So it's more convenient to have the meta file next to the file it configures.
I never said I had no solution, I just prefer to have a way to exclude the file.

**Comment by Ocramius** [comment]
Would suggest trying "exclude from classmap" or such - support probably to
be built for it

**Comment by VincentLanglet** [comment]
Thanks ! It works perfectly.

**Comment by Jean85** [comment]
Sorry for recovering an old issue, but I stumbled on another use case for this option: https://github.com/schmittjoh/serializer/issues/1381

The class is valid up to PHP 8.0, and the library wont drop the class due to BC. In the meantime my hack would be the same as https://github.com/schmittjoh/serializer/issues/1381#issuecomment-1000150120 (`rm`ing the file on the fly), but an exclusion on this side would be highly appreciated.

**Comment by maglnet** [comment]
Hi @Jean85 

maybe you can add the PHP 8.1 incompatible file to your `composer.json` - `autoload.exclude-from-classmap` config section?

https://github.com/maglnet/ComposerRequireChecker/blob/e737b9f17c6fdb40fd376e3e51a584df49636d8a/src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php#L52
